File: python-version/LC2234/maximum_beauty.py
class Solution:
    def maximumBeauty(self, flowers: list[int], new_flowers: int, target: int, full: int, partial: int) -> int:
        n = len(flowers)

        # 先假设全部种满，看看能剩下多少花
        left_flowers = new_flowers - target * n
        for i in range(n):  # 把原来有的花补上去
            flowers[i] = min(flowers[i], target)
            left_flowers += flowers[i]

        # 如果本来全都是满的，不需要种，直接返回
        if left_flowers == new_flowers:
            return full * n

        # 如果可以种满，从两种策略取其大，其一全部种满，其二留一个剩一朵，别的种满
        if left_flowers >= 0:
            return max(full * n, (n - 1) * full + (target - 1) * partial)

        flowers.sort()
        # 如果种不满，枚举能种满的后缀，多出来的补到前面去，选出最大的
        ans, pre_sum, j = 0, 0, 0
        # 枚举能种满的后缀，从1开始，0上面已经算过了
        for i in range(1, n + 1):
            # 撤销在i-1位置种的花
            left_flowers += target - flowers[i-1]
            if left_flowers < 0:  # 如果i位置后面还是种不满，继续撤销
                continue

            # 后面能种满的时候往前面种，种到flowers[j]
            while j < i and flowers[j] * j <= left_flowers + pre_sum:
                pre_sum += flowers[j]
                j += 1

            # 计算当前的美丽值
            avg = (left_flowers + pre_sum) // j
            beauty_value = avg * partial + full * (n-i)
            ans = max(ans, beauty_value)
        return ans

File: go-version/leetcode/LC2234/maximum_beauty.go
package LC2234

import "sort"

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/3/8 下午12:30
 * @FilePath: maximum_beauty
 * @Description: 前缀和验证博弈
 */

func maximumBeauty(flowers []int, newFlowers int64, target int, full int, partial int) int64 {
	n := len(flowers)

	//先把每个位置都种满target，看看能剩下多少花
	leftFlower := newFlowers - int64(target*n)
	for i := 0; i < n; i++ { //把已经有的加上去，这部分不用种
		flowers[i] = min(target, flowers[i])
		leftFlower += int64(flowers[i])
	}

	//如果本身全是满的，啥都不用种
	if leftFlower == newFlowers {
		return int64(full * n)
	}

	//如果需要种，但是可以全部种满，有两种方式，其一全部种满，其二留一个剩一朵，别的全种满
	if leftFlower >= 0 {
		return max(int64(partial*(target-1)+full*(n-1)), int64(full*n))
	}

	sort.Ints(flowers)
	//如果种不满，需要枚举能种满的后缀和种不满的前缀，选出最大的
	var ans, preSum int64
	j := 0
	//枚举后缀，从1开始，因为前面已经算过全部种满的情况了
	for i := 1; i <= n; i++ {
		//撤销在i-1位置种的花
		leftFlower += int64(target - flowers[i-1])
		if leftFlower < 0 { //撤销了还是不够种满后面的，需要继续撤销
			continue
		}

		//如果后面种满还有的剩，就往前面种，种到flowers[j]
		for j < i && int64(flowers[j]*j) <= leftFlower+preSum {
			preSum += int64(flowers[j])
			j++
		}
		//计算此时的美丽值
		avg := (preSum + leftFlower) / int64(j)
		beautyValue := int64(full*(n-i)) + int64(partial)*avg
		ans = max(ans, beautyValue)
	}
	return ans
}
